File: engine/client-csharp/Network/UdpClient.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using Zoit.Logger;
using Zoit.Protocol;

namespace Zoit.Network
{
    /// <summary>
    /// ZPP 프레임워크 전용 고성능 UDP 클라이언트
    /// </summary>
    public class UdpClient : IDisposable
    {
        private System.Net.Sockets.UdpClient _client;
        private IPEndPoint _remoteEndPoint;
        private CancellationTokenSource _cts;
        private bool _isStarted;

        public event Action<NetworkPacket> OnReceivePacket;

        public async Task<bool> StartAsync(string host, int port)
        {
            try
            {
                IPAddress[] addresses = await Dns.GetHostAddressesAsync(host);
                if (addresses.Length == 0) return false;

                _remoteEndPoint = new IPEndPoint(addresses[0], port);
                _client = new System.Net.Sockets.UdpClient();
                _client.Connect(_remoteEndPoint);
                _isStarted = true;
                _cts = new CancellationTokenSource();

                ClientLogger.Info($"[UDP] 클라이언트 시작: {host}:{port}");
                _ = ReceiveLoop(_cts.Token);
                return true;
            }
            catch (Exception ex)
            {
                ClientLogger.Error($"[UDP] 시작 실패: {ex.Message}");
                return false;
            }
        }

        public void Stop()
        {
            _isStarted = false;
            _cts?.Cancel();
            _client?.Close();
        }

        public void Send(uint command, byte[] body, uint senderIdx, uint version = 1)
        {
            if (!_isStarted) return;

            try
            {
                var header = new HeaderUDP
                {
                    Version = version,
                    Command = command,
                    Length = (uint)(body?.Length ?? 0),
                    Sender = senderIdx,
                    Error = 0
                };
                byte[] headerBytes = header.Encode();
                byte[] fullPacket = new byte[headerBytes.Length + (body?.Length ?? 0)];
                
                Buffer.BlockCopy(headerBytes, 0, fullPacket, 0, headerBytes.Length);
                if (body != null) Buffer.BlockCopy(body, 0, fullPacket, headerBytes.Length, body.Length);

                _client.Send(fullPacket, fullPacket.Length);
            }
            catch (Exception ex)
            {
                ClientLogger.Error($"[UDP] 패킷 전송 오류: {ex.Message}");
            }
        }

        private async Task ReceiveLoop(CancellationToken token)
        {
            try
            {
                while (!token.IsCancellationRequested && _isStarted)
                {
                    UdpReceiveResult result = await _client.ReceiveAsync();
                    byte[] data = result.Buffer;

                    if (data.Length < 20) continue;

                    var header = HeaderUDP.Decode(data);
                    byte[] body = null;
                    if (header.Length > 0 && data.Length >= 20 + header.Length)
                    {
                        body = new byte[header.Length];
                        Buffer.BlockCopy(data, 20, body, 0, (int)header.Length);
                    }

                    OnReceivePacket?.Invoke(new NetworkPacket(header.Command, body, header.Error, header.Sender));
                }
            }
            catch (Exception ex)
            {
                if (!token.IsCancellationRequested)
                {
                    ClientLogger.Warn($"[UDP] 수신 루프 종료: {ex.Message}");
                }
            }
        }

        public void Dispose()
        {
            Stop();
            _cts?.Dispose();
        }
    }
}

File: engine/client-python/network/udp_client.py
import asyncio
from ..protocol.header import HeaderUDP
from ..logger.logger import logger

class AsyncUdpClientProtocol(asyncio.DatagramProtocol):
    def __init__(self, on_receive=None):
        self.transport = None
        self.on_receive = on_receive

    def connection_made(self, transport):
        self.transport = transport

    def datagram_received(self, data, addr):
        if len(data) < 20:
            return

        try:
            header = HeaderUDP.decode(data[:20])
            body = data[20:] if header.length > 0 else b""
            
            if self.on_receive:
                if asyncio.iscoroutinefunction(self.on_receive):
                    asyncio.create_task(self.on_receive(header.command, body, header.sender, header.error))
                else:
                    self.on_receive(header.command, body, header.sender, header.error)
        except Exception as e:
            logger.error(f"[UDP] 패킷 처리 오류: {e}")

class AsyncUdpClient:
    def __init__(self, host, port, on_receive=None):
        self.host = host
        self.port = port
        self.on_receive = on_receive
        self.transport = None
        self.protocol = None

    async def start(self):
        try:
            loop = asyncio.get_event_loop()
            self.transport, self.protocol = await loop.create_datagram_endpoint(
                lambda: AsyncUdpClientProtocol(self.on_receive),
                remote_addr=(self.host, self.port)
            )
            logger.info(f"[UDP] 클라이언트 시작: {self.host}:{self.port}")
            return True
        except Exception as e:
            logger.error(f"[UDP] 시작 실패: {e}")
            return False

    async def send(self, command, body=None, sender_idx=0, version=1):
        if not self.transport:
            return
        
        try:
            body_bytes = body if body else b""
            header = HeaderUDP(version, command, len(body_bytes), sender_idx, 0)
            data = header.encode() + body_bytes
            self.transport.sendto(data)
        except Exception as e:
            logger.error(f"[UDP] 패킷 전송 오류: {e}")

    async def close(self):
        if self.transport:
            self.transport.close()
        logger.info("[UDP] 클라이언트 종료")
